scripts/compaction_bounds.zig

```diff
@@ -0,0 +1,128 @@
+                "id",
+                "debit_account_id",
+                "credit_account_id",
+                "user_data",
+                "pending_id",
+                "timeout",
+                "ledger",
+                "code",
+                "amount",
+                "timestamp",
+                "balance",
+            };
+            inline for (field_names) |field_name| {
+                const Value = comptime if (std.meta.eql(field_name, "timestamp"))
+                    // Object tree.
+                    Transfer
+                else if (std.meta.eql(field_name, "balance"))
+                    // Proposed balance tree.
+                    struct {
+                        account_id: u128,
+                        transfer_id: u128,
+                        debits_pending: u64,
+                        debits_posted: u64,
+                        credits_pending: u64,
+                        credits_posted: u64,
+                    }
+                else
+                    // Secondary index.
+                    IndexCompositeKeyType(@TypeOf(@field(@as(Transfer, undefined), field_name)));
+
+                const values_per_table = @divTrunc(config.lsm_table_size_max, @sizeOf(Value));
+                const table_count = try std.math.divCeil(usize, existing_transfer_count, values_per_table);
+
+                var level_count: usize = 0;
+                {
+                    var table_count_remaining: usize = table_count;
+                    var tables_per_level: usize = config.lsm_growth_factor;
+                    while (table_count_remaining > 0) {
+                        level_count += 1;
+                        table_count_remaining = table_count_remaining -| tables_per_level;
+                        tables_per_level *= config.lsm_growth_factor;
+                    }
+                }
+
+                // Write bandwidth for flushing the immutable table:
+                bandwidth_bytes_per_second +=
+                    // Total size of all values
+                    transfers_per_second * @sizeOf(Value) *
+                    // All trees except the object tree might have to write tombstones
+                    @as(usize, if (std.meta.eql(field_name, "timestamp")) 1 else 2);
+
+                // Read/write bandwidth for compaction:
+                const bandwidth_bytes_per_compaction =
+                    // Compaction touches:
+                    // * One table from level a.
+                    // * Up to config.lsm_growth_factor tables from level b.
+                    // TODO Compaction for immutable table to level 0 doesn't have to read immutable table from disk.
+                    (1 + config.lsm_growth_factor) *
+                    // The data in each table must be read from disk and written back out.
+                    // TODO This factor could actually be as high as 4 with tombstones.
+                    2 * config.lsm_table_size_max;
```

Comment:
We could drop the 2x multiplier here to keep everything in terms of write bandwidth only. Read bandwidth is typically higher and benefits from caching or pipelining.

Reply:
I've separated read and write bandwidth now.